### PR TITLE
Bluetooth API: add system and led state support

### PR DESCRIPTION
**Pull request by dexter93**
expose hooks for drivers to pick up

Implementation currently doing nothing, it's up to the bt drivers to implement it

Implementation example: https://github.com/SonixQMK/qmk_firmware/pull/292

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by ephb** [comment]
Is there any reason why this can not be merged or is only the review outstanding?

**Comment by fauxpark** [comment]
`bluetooth_send_system()` looks fine, the LED state stuff is a little trickier due to syncing issues.

**Comment by dexter93** [comment]
PR rebased. Keyboard specifics removed, waiting on CI to pick up any failures and will follow up with patches

**Comment by dexter93** [comment]
> `bluetooth_send_system()` looks fine, the LED state stuff is a little trickier due to syncing issues.
> 
> Also, there's a merge conflict.

syncing issues?

**Comment by fauxpark** [comment]
The LED state is not really a singular thing. Each keyboard interface has its own. This is not as much of an issue on Windows, where shift, lock and indicator state is shared across all connected keyboards - if you press Caps Lock, the host will tell all keyboards to turn the Caps Lock LED on, including *both* the 6KRO and NKRO interfaces of a QMK-based keyboard.

However macOS tracks this state separately for each connected device and interface. If you press Caps Lock, the LED turns on, but it is only for the 6KRO interface. If you then press `NK_TOGG`, the LED state will no longer match the OS, and even though the LED is on, typing will not give you uppercase letters. This problem extends to things like Bluetooth here.

**Comment by dexter93** [comment]
> The LED state is not really a singular thing. Each keyboard interface has its own. This is not as much of an issue on Windows, where shift, lock and indicator state is shared across all connected keyboards - if you press Caps Lock, the host will tell all keyboards to turn the Caps Lock LED on, including _both_ the 6KRO and NKRO interfaces of a QMK-based keyboard.
> 
> However macOS tracks this state separately for each connected device and interface. If you press Caps Lock, the LED turns on, but it is only for the 6KRO interface. If you then press `NK_TOGG`, the LED state will no longer match the OS, and even though the LED is on, typing will not give you uppercase letters. This problem extends to things like Bluetooth here.

I see, not surprised. However, that seems like a general issue not relevant to this PR. This one just enables state transmission over Bluetooth ( also, iirc most bluetooth HID modules do 6KRO - so, the syncing issue maybe not be triggered?)

**Comment by fauxpark** [comment]
> ( also, iirc most bluetooth HID modules do 6KRO - so, the syncing issue maybe not be triggered?)

Read what I said again. It's not about that.

**Comment by sigprof** [comment]
Is there any actual problem with the LED state introduced here?  `led_task()` gets invoked on every pass of the main loop; it will call `host_keyboard_leds()` and pick up any updates if the return value of `where_to_send()` had changed, so switching between USB and Bluetooth should not result in a stale LED state.  The problem with separate LED states between 6KRO and NKRO might be real, but it's probably out of scope of this PR.

Also for Bluetooth another use case which may be considered here is when USB and Bluetooth are connected to completely different hosts.

**Comment by fauxpark** [comment]
> Is there any actual problem with the LED state introduced here? led_task() gets invoked on every pass of the main loop; it will call host_keyboard_leds() and pick up any updates if the return value of where_to_send() had changed, so switching between USB and Bluetooth should not result in a stale LED state.

The host will only send LED updates when it needs to. It doesn't know when the keyboard switches between 6KRO/NKRO/Bluetooth/anything else.

**Comment by sigprof** [comment]
> The host will only send LED updates when it needs to. It doesn't know when the keyboard switches between 6KRO/NKRO/Bluetooth/anything else.

The USB protocol code would maintain its `keyboard_led_state` variable internally even when the Bluetooth connection is selected, so there should not be any sync issues with that.  The implementation of `bluetooth_led_state()` should also behave like that (if the BT connection is kept, the Bluetooth driver should still get the LED reports from the host and update its LED state accordingly, and if the connection is broken, the host would need to set the LED state again anyway).  Then switching between the USB and Bluetooth output would immediately update the keyboard LEDs to either the USB or Bluetooth state.

This leaves the 6KRO/NKRO case, but that part may be handled inside the USB driver (or even the Bluetooth driver, if some Bluetooth implementation would have NKRO support).

**Comment by dexter93** [comment]
broken userspace fixed on `keyboards/bioi`, remaining build failures are AVR size related on `spidey3` config, out of the context of this PR.

File: keyboards/bioi/morgan65/rules.mk [hunk 1]
```diff
@@ -16,7 +16,5 @@ AUDIO_ENABLE = no           # Audio output
 LTO_ENABLE = yes            # Reduce firmware size
 BLUETOOTH_ENABLE = yes
 
-VIA_ENABLE = yes            # VIA support should be enabled here due to the main() loop will be compiled first.
```
Comment (fauxpark) [inline keyboards/bioi/morgan65/rules.mk:19]: Good catch, VIA should not be enabled at the keyboard level.

Comment (dexter93) [reply keyboards/bioi/morgan65/rules.mk:19]: > Good catch, VIA should not be enabled at the keyboard level.

indeed, and the comment about it did not make any sense either. I should probably have moved the keyboard specifics in a new PR, but here they are. Any remaining merge blockers I should consider?

**Comment by fauxpark** [comment]
Merge conflicts.

Why don't we just get `bluetooth_send_system()` in for now, as that's more straightforward, and we can look at the LED state stuff after.

**Comment by dexter93** [comment]
> Merge conflicts.
> 
> Why don't we just get `bluetooth_send_system()` in for now, as that's more straightforward, and we can look at the LED state stuff after.

The merge conflicts appear to be on layout/userspace - will drop those since they're not really relevant to the PR.
I don't see why the LED state should matter at this level honestly - given it's a Mac issue in general with multiple keyboard interfaces. Perhaps it would be more appropriate to figure out a way to tackle that behavior for *every* scenario - and not just block the bt one?

**Comment by ephb** [comment]
Is the last word here that this is basically not wanted until the ```keyboard_led_state```` is reworked?

**Comment by fauxpark** [comment]
I'd prefer to get the `bluetooth_send_system()` part of this PR split out as that can go in immediately; I want to do some testing for the LED state stuff when I have time. That's the only thing "blocking" this PR.

**Pull request closed without merging**